syntax = "proto3";


message LogEntry {
    string cmd = 1;
    int64 term = 2;
}


message RequestVoteRequest {
    int64 term = 1;
    string candidateId = 2;
    int64 lastLogIndex = 3;
    int64 lastLogTerm = 4;
}


message RequestVoteResponse {
    int64 term = 1;
    bool voteGranted = 2;
}


message AppendEntriesRequest {
    int64 term = 1;
    string leaderId = 2;
    int64 prevLogIndex = 3;
    int64 prevLogTerm = 4;
    repeated LogEntry entries = 5;
    int64 leaderCommit = 6;
}


message AppendEntriesResponse {
    int64 term = 1;
    bool success = 2;
    int64 lastLogIndex = 3;
}


message Message {
    oneof Type {
        RequestVoteRequest request_vote = 1;
        RequestVoteResponse request_vote_response = 2;
        AppendEntriesRequest append_entries = 3;
        AppendEntriesResponse append_entries_response = 4;
    }
}
